pkg/controller/acmechallenges/sync.go

```diff
@@ -88,6 +88,16 @@ func (c *controller) Sync(ctx context.Context, ch *cmacme.Challenge) (err error)
 		return nil
 	}
 
+	// This finalizer ensures that the challenge is not garbage collected before
+	// cert-manager has a chance to clean up resources created for the
+	// challenge.
+	if setFinalizer := (&setFinalizerStep{
```

Comment:
What's the reason for creating a struct, instantiating it and then calling Run on it rather than having `finalizerRequired(ch)` followed by `addFinalizer(ch, c)`? It seems more complex to get your head around/understand is all :)

Reply:
Thanks for the review and sorry for the delayed reply. I was on holiday last week. 

The reason for the struct is that I had been thinking about how this Sync function could be made more readable and more easily testable.
The idea is that there will be a sequence of "steps" which each can be unit-tested in isolation and which can then be executed as follows with some wrapping which logs and creates an event (and maybe sets a status.condition) for each step, so that all the events can be consistent. 

Something like:


```go

	// The steps to be executed in order.
	//
	steps := []namedStep{
		{
			name:        "AddCleanupFinalizer",
			description: "Added finalizer to ensure that challenge resources are cleaned up",
			step: &addFinalizer{
				ch:               ch,
				challengeUpdater: c,
			},
		},
		{
			name:        "PresentChallenge",
			description: "Deploying resources for challenge",
			step: &presentChallenge{
				ch:     ch,
				solver: solver,
			},
		},
		{
			name:        "AcceptChallenge",
			description: "Accepting challenge with ACME server",
			step: &acceptChallenge{
				ch:     ch,
				solver: solver,
			},
		},
		{
			name:        "Cleanup",
			description: "Cleaning up the resources for challenge",
			step: &cleanupResources{
				ch:     ch,
				solver: solver,
			},
		},
	}
	for _, step := range steps {
		if !step.Required() {
			continue
		}
		reason := step.name
		message := step.description
		eventType := corev1.EventTypeNormal
		if err := action.Run(ctx); err != nil {
			reason = reason + "Failed"
			message = err.Error()
			eventType = corev1.EventTypeWarning
		}
		c.recorder.Event(ch, eventType, reason, message)
	}

```

But I am no longer sure that this quite fits the bill, so I'll simplify this PR to have a couple of unit-tested functions, as you suggest.